File: src/main/java/wraith/fabricaeexnihilo/json/other/WeightedListJson.java
package wraith.fabricaeexnihilo.json.other;

import com.google.gson.*;
import net.minecraft.block.Block;
import wraith.fabricaeexnihilo.api.crafting.WeightedList;
import wraith.fabricaeexnihilo.json.BaseJson;
import wraith.fabricaeexnihilo.json.basic.BlockJson;
import wraith.fabricaeexnihilo.util.RegistryUtils;

import java.lang.reflect.Type;
import java.util.HashMap;

public final class WeightedListJson extends BaseJson<WeightedList> {

    private WeightedListJson() {}

    public static final WeightedListJson INSTANCE = new WeightedListJson();

    @Override
    public WeightedList deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        HashMap<Block, Integer> map = new HashMap<>();
        for (var entry : json.getAsJsonObject().entrySet()) {
            map.put(BlockJson.INSTANCE.deserialize(new JsonPrimitive(entry.getKey()), BlockJson.INSTANCE.getTypeToken(), context), entry.getValue().getAsInt());
        }
        return new WeightedList(map);
    }

    @Override
    public JsonElement serialize(WeightedList src, Type typeOfSrc, JsonSerializationContext context) {
        var json = new JsonObject();
        src.getValues().forEach((k,v) -> json.add(RegistryUtils.getId(k).toString(), new JsonPrimitive(v)));
        return json;
    }

}

File: src/main/kotlin/exnihilofabrico/json/WeightedListJson.kt
package exnihilofabrico.json

import com.google.gson.*
import com.google.gson.reflect.TypeToken
import exnihilofabrico.api.crafting.WeightedList
import exnihilofabrico.util.getId
import java.lang.reflect.Type

object WeightedListJson: JsonDeserializer<WeightedList>, JsonSerializer<WeightedList> {
    val TYPE_TOKEN: Type = object: TypeToken<WeightedList>(){}.type
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): WeightedList {
        return WeightedList(
            json.asJsonObject.entrySet().map { (k, v) ->
                BlockJson.deserialize(JsonPrimitive(k), BLOCK_TYPE_TOKEN, context) to v.asInt
            }.toMap().toMutableMap()
        )
    }

    override fun serialize(src: WeightedList, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val json = JsonObject()
        src.values.forEach { (k,v) ->
            json.add(k.getId().toString(), JsonPrimitive(v))
        }
        return json
    }

}